wiki/python/ESTIMATOR_GUIDE.md

```diff
@@ -0,0 +1,709 @@
+np_arr = np.ones((10,))
+my_est.fit(np_arr) # This will load data into attributes
+
+# `my_cuml_array_` was set internally as a CumlArray. Externally, we can check the type
+print(type(my_est.my_cuml_array_)) # Output: Numpy (saved from the input of `fit`)
+
+# Calling fit again with cupy arrays, will have a similar effect
+my_est.fit(cp.ones((10,)))
+print(type(my_est.my_cuml_array_)) # Output: CuPy
+
+# Setting the `output_type` will change all descriptor properties
+# and ignore the input type
+my_est.output_type = "cudf"
+
+# Reading any of the attributes will convert the type lazily
+print(type(my_est.my_cuml_array_)) # Output: cudf
+
+# Setting the global_output_type, overrides the estimator output_type attribute
+with cuml.using_output_type("cupy"):
+  print(type(my_est.my_cuml_array_)) # Output: cupy
+
+# Once the global_output_type is restored, we return to the estimator output_type
+print(type(my_est.my_cuml_array_)) # Output: Cudf. Using a cached value!
+```
+
+The internal representation of `CumlArrayDescriptor` is a `CumlArrayDescriptorMeta` object. To inspect the internal representation, the attribute value must be directly accessed from the estimator's `__dict__` (`getattr` and `__getattr__` will perform the conversion). For example:
+
+```python
+my_est = TestEstimator()
+my_est.fit(cp.ones((10,)))
+
+# Access the CumlArrayDescriptorMeta value directly. No array conversion will occur
+print(my_est.__dict__["my_cuml_array_"])
+# Output: CumlArrayDescriptorMeta(input_type='cupy', values={'cuml': <cuml.common.array.CumlArray object at 0x7fd39174ae20>, 'numpy': array([ 0,  1,  1,  2,  2, -1, -1, ...
+
+# Values from CumlArrayDescriptorMeta can be specifically read
+print(my_est.__dict__["my_cuml_array_"].input_type)
+# Output: "cupy"
+
+# The input value can be accessed
+print(my_est.__dict__["my_cuml_array_"].get_input_value())
+# Output: CumlArray ...
+```
+
+#### Internal Array-Like Conversion
+
+Internally, developers can use the same `global_output_type` functionality to convert arrays as well. This has the same benefits of lazy conversion and caching as when descriptors are used externally. For example, we could add a `score()` function to `TestEstimator`:
+
+```python
+def score(self):
+
+   # Set the global output type to numpy
+   with cuml.using_output_type("numpy"):
+      # Accessing my_cuml_array_ will return a numpy array and
+      # the result can be returned directly
+      return np.sum(self.my_cuml_array_, axis=0)
+```
+
+### Estimator Functions
+
+In order for estimator functions to accept a wide variety of array types and correctly return the user's desired type, a fair amount of boiler plate code is involved. These common steps were often repetitive, inefficient, and incorrectly implemented.
+
+To assit with this, a set of decorators have been created to wrap estimator functions (and all `cuml` API functions as well) and perform the boiler plate actions automatically.
```

Comment:
I feel like this suggests we need decorators everywhere, but then after a while we recommend using automatic conversion. Maybe address that directly here, like "cuML provides two options to address this - (1) for many common patterns used in functions like fit, predict, transform, cuml.Base can automatically perform the data conversions as long as a method has type annotations, but (2) decorators are also provided to handle more advanced cases.

Reply:
Excellent suggestion. I struggled with organizing some of these pieces and I like this better. Will reorganize.

Reply:
See this comment: https://github.com/rapidsai/cuml/pull/3040#discussion_r518970546